src/query/sql/src/planner/optimizer/rule/rewrite/rule_fold_constant.rs

```diff
@@ -0,0 +1,184 @@
+                // EvalScalar
+                //  \
+                //   *
+                SExpr::create_unary(
+                    PatternPlan {
+                        plan_type: RelOp::EvalScalar,
+                    }
+                    .into(),
+                    SExpr::create_leaf(
+                        PatternPlan {
+                            plan_type: RelOp::Pattern,
+                        }
+                        .into(),
+                    ),
+                ),
+                // ProjectSet
+                //  \
+                //   *
+                SExpr::create_unary(
+                    PatternPlan {
+                        plan_type: RelOp::ProjectSet,
+                    }
+                    .into(),
+                    SExpr::create_leaf(
+                        PatternPlan {
+                            plan_type: RelOp::Pattern,
+                        }
+                        .into(),
+                    ),
+                ),
+                // Exchange
+                //  \
+                //   *
+                SExpr::create_unary(
+                    PatternPlan {
+                        plan_type: RelOp::Exchange,
+                    }
+                    .into(),
+                    SExpr::create_leaf(
+                        PatternPlan {
+                            plan_type: RelOp::Pattern,
+                        }
+                        .into(),
+                    ),
+                ),
+                // Join
+                //  \
+                //   *
+                SExpr::create_unary(
+                    PatternPlan {
+                        plan_type: RelOp::Join,
+                    }
+                    .into(),
+                    SExpr::create_leaf(
+                        PatternPlan {
+                            plan_type: RelOp::Pattern,
+                        }
+                        .into(),
+                    ),
+                ),
+            ],
+            func_ctx,
+        }
+    }
+
+    fn fold_constant(&self, scalar: &ScalarExpr) -> Result<ScalarExpr> {
+        let expr = scalar.as_expr()?;
+        let (new_expr, _) = ConstantFolder::fold(&expr, &self.func_ctx, &BUILTIN_FUNCTIONS);
+        let new_scalar = ScalarExpr::from_expr(&new_expr)?;
+        Ok(new_scalar)
+    }
+}
+
+impl Rule for RuleFoldConstant {
+    fn id(&self) -> RuleID {
+        self.id
+    }
+
+    fn apply(
+        &self,
+        s_expr: &SExpr,
+        state: &mut crate::optimizer::rule::TransformResult,
+    ) -> Result<()> {
+        let mut new_plan = s_expr.plan().clone();
+        match &mut new_plan {
+            RelOperator::EvalScalar(eval_scalar) => {
+                for scalar in eval_scalar.items.iter_mut() {
+                    scalar.scalar = self.fold_constant(&scalar.scalar)?;
+                }
+            }
+            RelOperator::Filter(filter) => {
+                for predicate in filter.predicates.iter_mut() {
```

Comment:
For `RelOperator::Filter`, could combine with `and` operator and `cast_expr_to_non_null_boolean`.

Then apply `fold_constant`, thus we don't need do it again in physical_plan_builder.

Reply:
It relies on the pre-split conjunctions to perform some pushdown optimization, I think it's better not to combine the conjunctions here into a single `ScalarExpr`.